### DISCUSSION TITLE
XRechnung/Zugferd support

### DISCUSSION THREAD

**Discussion by wobalrz** [kimai/kimai#4575] [General]
According to German regulation, electronic invoices following EU standard EN16931 will be mandatory by Jan 1, 2025 for receiving/accepting invoices; after a transition period it will require companies to issue invoices electronically  (currently it's mandatory only for invoices to contracting public agencies, but this is to be extended to all B2B invoices, and also affects freelancers):

https://www.datev.de/web/de/aktuelles/gesetzliche-themen/e-rechnung/e-rechnung-neue-entwicklungen-und-gesetzliche-regelungen/e-rechnung-entwicklung-fuer-deutschland/

Are there any plans to support XRechnung in kimai?
Though it appears to be a purely German requirement right now, it's based on EU regulation and thus likely to be relevant to all EU countries sooner or later, and at least for Germans it'll be a "must have" soon.

**Comment by kevinpapst** [comment]
> Ab dem 01.01.2027 müssen Unternehmen mit einem Vorjahresumsatz von mehr als 800.000 Euro im B2B-Bereich E-Rechnungen versenden. Unternehmen mit einem Vorjahresumsatz von weniger als 800.000 Euro dürfen bis zum 31.12.2027 noch sonstige Rechnungen (Papier, PDF etc.) versenden.

Not as urgent as it sounds.
If your company wants to sponsor development, I am open for talks.

You can also vote for https://github.com/kimai/kimai/issues/1396

**Reply by wobalrz** [reply]
I don't fully agree - larger companies in particular want to get rid of PDF invoices and OCR solutions (which still break a fully digital workflow all too often), and require contractors to comply with XRechnung  via their terms&conditions well before the ultimate legal deadline - as German public agencies already do.

Just wanted to know whether that's already on the agenda.

**Reply by kevinpapst** [reply]
Whenever such a dedicated topic around earning and invoicing comes up, all I can ask myself is: why shall I cover the expenses for the development of these features?

Users from Canada asks me to support multiple tax rates. 
Users from certain districts in Italy want them as well.
Swiss users want their QR codes on invoices.
Users from India have even more interesting topics.
Users from Germany now want XRechnung (who decided on that insanely stupid name?).

You can vote for Zugferd and you can contact me via email if you want to throw money at the topic.
You can open feature requests for these issues. 
But this is a discussion, so you get discussion feedback.

**Reply by wobalrz** [reply]
Isn't it the nature of a discussion to share views on issues such as why a feature may be (not) relevant for the broader community!? 

Got it...

**Reply by kevinpapst** [reply]
I shared my view and you closed the issue 🤷 that doesn't make sense at all. 

What you are asking: is maintainer is willing to add such a feature to Kimai?
The maintainer answered: yes, but somebody needs to sponsor such a feature. 

> Isn't it the nature of a discussion to share views on issues such as why a feature may be (not) relevant for the broader community!?

Obviously that feature is relevant. But if you are unhappy with the answer and immediately close the discussion, then you are clearly not here for the discussion itself, but rather for the free beer. Let me tell you: you misunderstand the term "free" in "free & open source software".

If you need a dedicated invoicing software, it will cost money as well. 
So closing the discussion and not giving other the options to express their interest is short-sighted.

**Comment by pronego-mls** [comment]
Hello,
Just to reopen this issue again as it is still relevant, as it will be mandatory even for small companies with 1.1.2028.

What is the current state here?
@kevinpapst: How much would you estimate for the implementation of XRechnung+ZUGFeRD?

If we'd have some number, we could coordinate next steps in terms of funding, e.g. via crowd, etc.
Maybe other interested parties would like to comment on this, too...

**Reply by kevinpapst** [reply]
Get in touch via support@kimai.org - I am actively working on it

**Reply by pronego-mls** [reply]
That's great news.
Is there anything I could support you with?
And do you already have planned a release date?

**Reply by kevinpapst** [reply]
Get in touch via [support@kimai.org](mailto:support@kimai.org)

**Reply by BeckeBauer** [reply]
> If we'd have some number, we could coordinate next steps in terms of funding, e.g. via crowd, etc. Maybe other interested parties would like to comment on this, too...

Suggested this as well since it seems that many companies are waiting for this feature but my impressions is that crowed funding is not liked...

**Reply by kevinpapst** [reply]
No, I am not using Crowdfunding. 
Everyone who got in touch in the past will get an update next week, regarding testing.
It am almost done with the implementation.

The feature will be shipped as plugin, as this is quite complex and very specific for the EU, especially Germany and France.

**Comment by facturx-engine** [comment]
@kevinpapst Massive respect for taking on the ZUGFeRD/XRechnung implementation natively as a plugin. 

Having wrestled with developing an e-invoicing engine recently, I completely agree with you: the sheer volume of EU-specific edge cases (like rounding margins on nested tax categories, let alone the nightmare of strict PDF/A-3 XMP metadata injections in PHP) makes this a brutal feature to build and maintain for a single developer.

Just wanted to share a resource in case it saves you (or anyone testing the plugin) some sleepless nights battling PDF libraries or strict government validators. 

I open-sourced **Factur-X Engine** recently. It’s a stateless Docker REST API that completely abstracts the EN 16931 validation and PDF/A-3 packaging. 

If the native PHP PDF generation ever becomes too brittle against the German KoSIT validators, or if you need a deterministic backend to bounce the generated XML against during testing, feel free to use it. 

Thanks for pushing Kimai forward for the EU ecosystem, looking forward to the plugin release!

**Reply by kevinpapst** [reply]
Come on, spreading links in neighbour repositories you don't use is not necessary.
You don't care about the plugin, you care about spreading the word of your tool. 
Which I kinda understand and you do it actually in a very friendly way, but still...
